build.assets/robotest/Makefile

```diff
@@ -0,0 +1,326 @@
+		--build-arg GID=$(shell id -g) \
+		--iidfile $(ROBOTEST_DOCKER_IMAGE_IIDFILE) \
+		--tag $(ROBOTEST_DOCKER_IMAGE) .
+
+.PHONY: images
+images: ## Create all images necessary to run robotest CI tests.
+images: image-robotest images-upgrade image-opscenter
+
+.PHONY: image-robotest
+image-robotest: ## Create robotest cluster image using the current tele.
+image-robotest: $(ROBOTEST_IMAGE)
+
+$(ROBOTEST_IMAGE): $(TELE_OUT) $(ROBOTEST_IMAGE_SRC) | $(ROBOTEST_IMAGEDIR) $(ROBOTEST_BUILD_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_BUILD_TMP)))
+	$(eval TGT := $(TMP)/robotest-$(GRAVITY_VERSION).tar)
+	$(TELE_OUT) build \
+		$(ROBOTEST_APP_YAML) \
+		--state-dir=$(PACKAGES_DIR) \
+		--version $(GRAVITY_VERSION) \
+		--output=$(TGT)
+	mv $(TGT) $(ROBOTEST_IMAGE)
+	rm -rf $(TMP)
+
+.PHONY: images-upgrade
+images-upgrade: ## Create/fetch robotest cluster images needed for upgrade testing.
+images-upgrade:
+	$(MAKE) -j $(ROBOTEST_UPGRADE_IMAGES)
+
+.PHONY: image-opscenter
+image-opscenter: ## Move opscenter.tar cluster image into place for test execution.
+image-opscenter: $(ROBOTEST_OPSCENTER_IMAGE)
+
+$(ROBOTEST_OPSCENTER_IMAGE): $(OPSCENTER_OUT) | $(ROBOTEST_IMAGEDIR)
+	cp $(OPSCENTER_OUT) $(ROBOTEST_OPSCENTER_IMAGE)
+
+# The following rule populates the build specific docker image mount
+# from cached images. Fulfills the $(ROBOTEST_UPGRADE_IMAGES) targets.
+# .PRECIOUS because these need to be present through the run target.
+.PRECIOUS: $(ROBOTEST_IMAGEDIR)/robotest-%.tar
+$(ROBOTEST_IMAGEDIR)/robotest-%.tar: $(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar | $(ROBOTEST_IMAGEDIR) $(ROBOTEST_BUILD_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_BUILD_TMP)))
+	$(eval ATOMIC := $(TMP)/robotest-$*.tar)
+	cp $< $(ATOMIC)
+	mv $(ATOMIC) $@
+	rm -rf $(TMP)
+
+# The following target is a pattern rule for populating the upgrade image cache
+# Fulfills the $(ROBOTEST_UPGRADE_CACHE_IMAGES) targets.
+# .PRECIOUS because it is a cache shared across versions.
+.PRECIOUS: $(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar
+$(ROBOTEST_UPGRADE_IMAGE_CACHE)/robotest-%.tar: $(ROBOTEST_TELE_CACHE)/tele-% $(ROBOTEST_UPGRADE_BASE_IMAGE_SRC) | $(ROBOTEST_UPGRADE_IMAGE_CACHE) $(ROBOTEST_CACHE_TMP)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_CACHE_TMP)))
+	$(eval TGT := $(TMP)/robotest-$*.tar)
+	$< build \
+		$(ROBOTEST_UPGRADE_BASE_APP_YAML) \
+		--state-dir $(ROBOTEST_GRAVITY_PKG_CACHE) \
+		--version $* \
+		--output=$(TGT)
+	mv $(TGT) $@
+	rm -rf $(TMP)
+
+
+# The following target is a pattern rule for populating the tele binary cache.
+# Fulfills the $(ROBOTEST_TELE_BINARIES) targets.
+# .PRECIOUS because it is a cache shared across branches/builds.
+.PRECIOUS: $(ROBOTEST_TELE_CACHE)/tele-%
+$(ROBOTEST_TELE_CACHE)/tele-%: | $(ROBOTEST_CACHE_TMP) $(ROBOTEST_TELE_CACHE)
+	$(eval TMP := $(shell mktemp -d -p $(ROBOTEST_CACHE_TMP)))
+ifdef ENTERPRISE
+	wget --quiet https://get.gravitational.io/telekube/bin/$*/linux/x86_64/tele --output-document $(TMP)/tele
```

Comment:
Can we find a way to avoid placing enterprise-specific URLs in the OSS code? For example, instead of setting `ENTERPRISE` variable in the enterprise code, set `TELE_URL` or something like this and use it here.

Reply:
The issue here is that OSS gravity doesn't provide a direct link to download tele that I've found - we need to download it as part of a tarball, unpack and then copy just the `tele` binary.  Whereas for enterprise we need to download just the binary, or authenticate to get.gravitational.io to get a release tarball.  I haven't been able to find a common download path.

If we wanted to properly abstract this difference in behavior, we'd need to either:
* Find a common download path.  Two unauthenticated release tarballs or two unauthenticated raw binaries.
* Declare & call a function pointer, if such a thing is even possible in a Makefile. This seems like an poor choice from a maintenance perceptive.
* Do some sort of `sed`-esque replacement of a `TELE_URL` variable to plug in the variety of upgrade-from versions.

Let me know what you think.  The value of keeping enterprise stuff out of OSS stuff is clear to me, but the path to doing so is not.

Reply:
I can imagine having some e-specific make functions stored in e/ which could be conditionally included with:
```
-include e/makefile.conf
```
Then we can factor the tele pull process into a function that accepts the version and output path:
```
define pull_tele
	wget --quiet https/path/to/telekube/bin/$1/linux/x86_64/tele --output-document $2
	chmod u+x $2
endef
```

Reply:
Thanks @a-palchikov.  I'll run with your suggestion.

Reply:
I've got a working version of this change in place as of c91653e.

I use a shell script instead of a included Makefile to accommodate tmpdir cleanup per https://github.com/gravitational/gravity/pull/2133#discussion_r489093866.

I want to make one more structural change in this area before calling it done.  That is:

Migrate all content from `build.assets/robotest` to `assets/robotest`.  This will give a nice parity between the enterprise `assets/robotest` and the oss `assets/robotest`.

Reply:
Take a look at `assets/robotest/download_tele.sh` in the two repos and see if the new implementation suits your fancy.